### ISSUE TITLE
Make random.sample's k argument optional

### ISSUE THREAD

**Issue by Gouvernathor** [python/cpython#96545]
# Feature or enhancement

The random.sample function takes a k argument, I want to make it default to the length of the sequence.

# Pitch

It would make a copy-shuffle (as opposed to the in-place shuffle done by random.shuffle) just as simple as one function call. It would also allow easier of `random.sample` in a functional programming fashion, by avoiding the need to create a variable containing the population variable prior to calling the function.
In other words, `random.sample` would be to `random.shuffle` what `sorted` is to `str.sort`.

# Previous discussion

https://discuss.python.org/t/making-random-samples-k-argument-optional/

**Comment by rhettinger** [comment]
Thanks for the suggestion, but I don't think it is the right thing for most users. The proposed default is inconsistent with `choices()` where the default is `k=1`.  Also, it a large population is common. Having the default be the full population size would be an unpleasant surprise.

It is better to stick to the intended purpose of making a small sample from a large population and avoid feature creep into the use case for shuffling everything (which is only reasonable to do for small population sizes).

Besides the issues of an inconsistent default, a hazardous default, and of feature creep, I think this is also a case where "explicit is better than implicit".  The word "sample" generally isn't taken to mean "shuffle everything", so it is better for a user to be explicit and write `sample(cards, k=len(cards))` to indicate that they are sampling the whole deck rather than a portion of it.

Thanks for the suggestion, but I think it is ill-advised.   

One last thought is that the use case is not typically what is desired. While the proposal does shorten a one-liner by a few characters, it tends to steer users away from an in-place shuffle which is often the better approach. We don't want to make the worse way to do it more attractive than the best way to do it.

**Comment by pochmann** [comment]
I've been wishing for a `shuffled` function quite often. I'd much prefer that over using `sample` without `k`, though, as the intention is really not sampling but shuffling. While Python's `random.sample` does incorporate random order, I don't think that's an inherent part of sampling in general. In fact, both [Wikipedia](https://en.m.wikipedia.org/wiki/Sampling_(statistics)) and [MathWorld](https://mathworld.wolfram.com/Sample.html) say sampling gives a *subset* of a population, and sets are unordered. Also, `shuffle`/`shuffled` would match `sort`/`sorted` more nicely. And you yourself also describe it in terms of shuffling ("make a copy-shuffle").

**Comment by Gouvernathor** [comment]
Let's take this back to discuss, if you will.

**Comment by rhettinger** [comment]
One extra note:  [numpy.random_sample()](https://numpy.org/doc/stable/reference/random/generated/numpy.random.random_sample.html#numpy.random.random_sample) has a default sample size of one.

**Comment by Gouvernathor** [comment]
It's a multi-dimentional extension of random.random(), it has nothing to do wth random.sample apart from the name (and the use of randomness).

**Comment by pochmann** [comment]
Grrr... the discuss was closed, by someone saying to be considerate of people's time, thereby **wasting** my time I spent writing a comment I now can't post there anymore. Alright, lesson learned, I won't participate in discuss anymore. Anyway, I don't want it to go to waste, so exceptionally I post it here and then I'm out.

[quote="Gouvernathor, post:29, topic:18742, username:Gouvernathor"]
need actual sequences in order for the pseudo-randomization to be reliable and deterministic.
[/quote]

I think the main reason choice/choices/sample want sequences is efficiency. If you want to pick one/some random element(s) from a dict, how do you do that efficiently? I don't think you can. At least not without bypassing its interface and directly messing around in its internal data structures (which would be *really* messy).

For shuffling, that efficiency concern doesn't exist, as it shuffles the whole thing anyway. Not sure what you mean with reliability. Determinism is an argument, but I'm an adult. If I don't need reproducibility, I don't need it. I'd prefer being able to get a shuffled version of any iterable.

**Comment by Gouvernathor** [comment]
Reproducibility is the point of the bookkeeping functions of the random module. It's a guarantee by the module that if you set a certain seed, and pass a certain value to a certain random function, the result will first be unpredictable, but then always, always be the same.
Using a dict is somehow safe, since their ordering is documented since some versions ago. But sets however are not. When doing `list(someset)`, there's no information at all about the order of the result. So if I call `random.sample(lsomeset, k)`, in the versions when that was possible, it would unpredictably turn the set into a list, making the whole operation undeterministic, contradictory to the module's promise.
In the current version, it basically says that if *you* don't care about ordering and determinism and actually want something as close to truly random as it gets, then it's your prerogative, but it's *your* responsibility to call `list()` or `tuple()` on your set or dict or non-sequence iterable.
It may not be what _you_ want, but some people are happy to be able to reliably and deterministically shuffle their lists. I know I am. I used it in a project of mine and I was very happy about it.

**Comment by pochmann** [comment]
Where can I see that guarantee?

**Comment by Gouvernathor** [comment]
There : https://docs.python.org/3/library/random.html#notes-on-reproducibility

**Comment by serhiy-storchaka** [comment]
The topic on [discuss.python.org](https://discuss.python.org/t/making-random-samples-k-argument-optional/) was renamed for adding `random.sampled`. I want to remind that this idea was already proposed and rejected 3 times.

* https://github.com/python/cpython/issues/70581
* https://github.com/python/cpython/issues/72151
* https://github.com/python/cpython/issues/85629

**Comment by Gouvernathor** [comment]
Yep, rejected three times after no discussion at all, and now a fourth after someone throw a bogus accusation which forcefully closed the argument. What a great marketplace of ideas.

**Comment by pochmann** [comment]
@Gouvernathor That doesn't guarantee what you say it guarantees. It only guarantees something about the `random()` function, not about any other.
